Description: When using Drag'n'Drop from Places menu, install a link launcher (.desktop file) instead of copying the entire directory.
Author: Marcel Stimberg
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=357934
Bug-Ubuntu: https://launchpad.net/bugs/62529 

Index: gnome-panel-2.30.0/gnome-panel/menu.c
===================================================================
--- gnome-panel-2.30.0.orig/gnome-panel/menu.c	2010-03-09 00:07:06.000000000 +1100
+++ gnome-panel-2.30.0/gnome-panel/menu.c	2010-06-03 13:57:54.293946027 +1000
@@ -1213,7 +1213,8 @@
 void
 setup_uri_drag (GtkWidget  *menuitem,
 		const char *uri,
-		const char *icon)
+		const char *icon,
+		GdkDragAction action)
 {
 	static GtkTargetEntry menu_item_targets[] = {
 		{ "text/uri-list", 0, 0 }
@@ -1225,7 +1226,7 @@
 	gtk_drag_source_set (menuitem,
 			     GDK_BUTTON1_MASK|GDK_BUTTON2_MASK,
 			     menu_item_targets, 1,
-			     GDK_ACTION_COPY);
+			     action);
 
 	if (icon != NULL)
 		gtk_drag_source_set_icon_name (menuitem, icon);
Index: gnome-panel-2.30.0/gnome-panel/menu.h
===================================================================
--- gnome-panel-2.30.0.orig/gnome-panel/menu.h	2010-01-14 14:26:27.000000000 +1100
+++ gnome-panel-2.30.0/gnome-panel/menu.h	2010-06-03 13:57:54.293946027 +1000
@@ -51,7 +51,8 @@
 					    PanelActionButtonType  type);
 void            setup_uri_drag             (GtkWidget  *menuitem,
 					    const char *uri,
-					    const char *icon);
+					    const char *icon,
+						GdkDragAction action);
 
 GtkWidget *	panel_create_menu              (void);
 
Index: gnome-panel-2.30.0/gnome-panel/panel-menu-items.c
===================================================================
--- gnome-panel-2.30.0.orig/gnome-panel/panel-menu-items.c	2010-06-03 13:57:49.626446947 +1000
+++ gnome-panel-2.30.0/gnome-panel/panel-menu-items.c	2010-06-03 13:57:54.293946027 +1000
@@ -261,7 +261,7 @@
 
 	uri = g_filename_to_uri (full_path, NULL, NULL);
 
-	setup_uri_drag (item, uri, icon);
+	setup_uri_drag (item, uri, icon, GDK_ACTION_COPY);
 	g_free (uri);
 
 	g_key_file_free (key_file);
@@ -308,7 +308,8 @@
 	g_signal_connect (G_OBJECT (item), "button_press_event",
 			  G_CALLBACK (menu_dummy_button_press_event), NULL);
 
-	setup_uri_drag (item, uri, icon_name);
+	if (g_str_has_prefix (uri, "file:")) /*Links only work for local files*/
+		setup_uri_drag (item, uri, icon_name, GDK_ACTION_LINK);
 }
 
 static GtkWidget *
